### PR TITLE
Add assertion that terminate clio

### LINKED ISSUE #893: Improve asserts in the code
- Create a macro `ASSERT()` to log enough information and close clio
- For each `assert()` in our code:
  - replace it with the `ASSERT()` macro if it checks invariants violation
  - replace it with a proper exception throwing to recover at the right level of the application (probably at request processing level)
  - if this will be done after #888 we should report to prometheus recoverable errors

### LINKED ISSUE #929: Use Boost.Stacktrace or std equivalent
Add stacktrace to crash log.

See https://www.boost.org/doc/libs/1_82_0/doc/html/stacktrace.html 
and https://en.cppreference.com/w/cpp/utility/basic_stacktrace

### PR DESCRIPTION
**Pull request by kuznetsss**
Fixes #893.

Also added termination handler to print backtrace on crash, so fixes #929.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -56,11 +57,14 @@ target_link_libraries (clio
...
   PUBLIC Boost::log
   PUBLIC Boost::log_setup
+  PUBLIC Boost::stacktrace_backtrace
   PUBLIC cassandra-cpp-driver::cassandra-cpp-driver
   PUBLIC fmt::fmt
   PUBLIC OpenSSL::Crypto
   PUBLIC OpenSSL::SSL
   PUBLIC xrpl::libxrpl
+  PUBLIC dl
+  PUBLIC libbacktrace::libbacktrace
```
Comment (godexsoft) [inline CMakeLists.txt:67]: Do we still need these two even with Boost.Stacktrace?

File: conanfile.py [hunk 2]
```diff
@@ -27,6 +27,7 @@ class Clio(ConanFile):
         'grpc/1.50.1',
         'openssl/1.1.1u',
         'xrpl/2.0.0-rc1',
+        'libbacktrace/cci.20210118'
```
Comment (godexsoft) [inline conanfile.py:30]: Is this needed with Boost.Stacktrace still?

File: src/data/cassandra/impl/ExecutionStrategy.h [hunk 3]
```diff
@@ -455,10 +466,7 @@ class DefaultExecutionStrategy {
     decrementOutstandingRequestCount()
     {
         // sanity check
-        if (numWriteRequestsOutstanding_ == 0) {
-            assert(false);
-            throw std::runtime_error("decrementing num outstanding below 0");
-        }
+        ASSERT(numWriteRequestsOutstanding_ > 0, "decrementing num outstanding below 0");
```
Comment (godexsoft) [inline src/data/cassandra/impl/ExecutionStrategy.h]: Let's keep the messages in the same format. i.e. capitalised 👍

File: src/etl/Source.h [hunk 4]
```diff
@@ -794,7 +794,7 @@ class SourceImpl : public Source {
                 uint32_t const sequence = std::stoll(minAndMax[0]);
                 pairs.emplace_back(sequence, sequence);
             } else {
-                assert(minAndMax.size() == 2);
+                ASSERT(minAndMax.size() == 2, "Min and max should be of size 2. Got size ={}", minAndMax.size());
```
Comment (godexsoft) [inline src/etl/Source.h]: Please add a space between `=` and `{}` 👍

File: src/etl/impl/AsyncData.h [hunk 5]
```diff
@@ -60,7 +61,12 @@ class AsyncCallData {
                           << " . prefix = " << ripple::strHex(std::string(1, prefix))
                           << " . nextPrefix_ = " << ripple::strHex(std::string(1, nextPrefix_));
 
-        assert(nextPrefix_ > prefix || nextPrefix_ == 0x00);
+        ASSERT(
+            nextPrefix_ > prefix || nextPrefix_ == 0x00,
+            "nextPrefix_ must be greater than prefix. Got: nextPrefix_ = {}, prefix = {} ",
```
Comment (godexsoft) [inline src/etl/impl/AsyncData.h]: I think it's better to say "Next prefix must be greater than current prefix." to keep the formatting. 
Also let's remove the last space 🔢

File: src/etl/impl/LedgerLoader.h [hunk 6]
```diff
@@ -189,50 +189,49 @@ class LedgerLoader {
...
-                                    assert(succ);
-                                    if (succ->key == cur->key) {
-                                        LOG(log_.debug()) << "Writing book successor = " << ripple::strHex(base)
-                                                          << " - " << ripple::strHex(cur->key);
-
-                                        backend_->writeSuccessor(
-                                            uint256ToString(base), sequence, uint256ToString(cur->key)
-                                        );
-                                    }
+                auto seconds = ::util::timed<std::chrono::seconds>([this, edgeKeys = &edgeKeys, sequence, &numWrites](
```
Comment (godexsoft) [inline src/etl/impl/LedgerLoader.h]: If you remove the unnecessary `()` from the lambda it may format better 👍

File: src/main/Main.cpp [hunk 7]
```diff
@@ -218,4 +220,8 @@ try {
     return EXIT_SUCCESS;
 } catch (std::exception const& e) {
     LOG(LogService::fatal()) << "Exit on exception: " << e.what();
+    return EXIT_FAILURE;
+} catch (...) {
```
Comment (godexsoft) [inline src/main/Main.cpp:224]: Good idea 👍

File: src/util/Assert.h [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+#include <string_view>
+
+#include <boost/stacktrace.hpp>
+#include <fmt/format.h>
+
+namespace util {
+
+template <typename... Args>
+constexpr void
+assert_impl(
```
Comment (godexsoft) [inline src/util/Assert.h]: CamelCase please 🔤

File: src/util/Assert.h [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    char const* expression,
+    bool const condition,
+    fmt::format_string<Args...> format,
+    Args&&... args
+)
+{
+    if (!condition) {
+        fmt::println(stderr, "Assertion '{}' failed at {}:{}:", expression, location.file_name(), location.line());
+        fmt::println(stderr, format, std::forward<Args>(args)...);
+        std::cerr << "Stacktrace:\n" << boost::stacktrace::stacktrace() << std::endl;
```
Comment (godexsoft) [inline src/util/Assert.h]: Why not use the same `println` function for this? In `libfmt` you can use everything that's compatible with streams if you include `fmt/std.h` if i remember correctly.

File: unittests/Main.cpp [hunk 10]
```diff
@@ -16,13 +16,15 @@
     OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
 //==============================================================================
+#include <util/TerminationHandler.h>
```
Comment (godexsoft) [inline unittests/Main.cpp:20]: There should always be a line of space after the `//==` line.

**Review by godexsoft — CHANGES_REQUESTED** [review]
Very cool 🔥 Let's fix the nitpicks; otherwise looks very useful 🙇

File: src/util/Assert.h [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#pragma once
+
+#include <util/SourceLocation.h>
+
+#include <iostream>
+#include <string_view>
```
Comment (cindyyan317) [inline src/util/Assert.h]: do we need this header?

File: CMakeLists.txt (hunk 1 above)
Comment (kuznetsss) [reply CMakeLists.txt:67]: Yes, Boost.stacktrace requires these two. https://www.boost.org/doc/libs/1_83_0/doc/html/stacktrace/configuration_and_build.html

File: conanfile.py (hunk 2 above)
Comment (kuznetsss) [reply conanfile.py:30]: As I understand Boost.Stacktrace is a wrapper around this library (if Boost.stacktrace is used in  backtrace mode).

File: src/util/Assert.h (hunk 9 above)
Comment (kuznetsss) [reply src/util/Assert.h]: Still one more line is needed to define `template<> fmt::formatter<boost::stacktrace::stacktrace>`.

**Review by godexsoft — DISMISSED** [review]
LGTM 🚀

**Review by godexsoft — APPROVED** [review]

**Pull request merged**